integration: use systemd when $DOCKER_SYSTEMD is set

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Added systemd in the integration testing image.
This will help testing rootless and systemd stuff.

Fix #40492

**- How I did it**
Added a variant of `hack/dind` script:`hack/dind-systemd`.

This script executes the arguments as a systemd service in foreground with tty connected, and then call `systemctl exit $EXIT_STATUS` to shutdown the systemd process with the expected exit code.


**- How to verify it**
* `DOCKER_SYSTEMD=1 make test-integration`

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
:penguin:

AkihiroSuda committed:
integration: use systemd when $DOCKER_SYSTEMD is set

Fix #40492

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented on hack/dind:
```diff
@@ -26,8 +26,33 @@ if ! mountpoint -q /tmp; then
+After=hack-dind.target
+
+[Service]
+Type=oneshot
```
This seems making `systemctl status` to be "initializing". Not sure how to make the status "running".

AkihiroSuda replied:
Removing Type=oneshot solved the issue

thaJeztah commented:
This script is more widely used than just in our CI; wondering if we should have a separate script _just_ for CI if this is needed

AkihiroSuda commented:
Testing should be possible on a laptop as well?

thaJeztah commented:
Right, but it's also used in the `docker` images on Docker Hub for example; https://github.com/docker-library/docker/blob/7db7eccb7b7507b1a6ba3ef81c6f37b16149cc55/Dockerfile-dind.template#L33-L38

Also not sure

- if we always want systemd to be used?
- if instead of `cat`-ing the unit file, we should just add a unit file to this repo, and `COPY` it into the container

AkihiroSuda commented:
Updated not to use systemd when systemd is not installed, but I think `docker` images on Docker Hub should fork the code.

> if we always want systemd to be used?

I think we can use systemd always.

Also, even for rootful with cgroupfs driver, systemd might be useful for simplifying `.integration-daemon-{start,stop}` stuff.

> if instead of cat-ing the unit file, we should just add a unit file to this repo, and COPY it into the container

No, the file needs to be autogenerated (for command args, workdir, and env vars)

AkihiroSuda commented:
Interestingly, this PR seems to have digged up some broken tests that hadn't been in CI.

```
[2020-02-10T21:34:56.998Z] --- FAIL: TestCgroupDriverSystemdMemoryLimit (1.76s)

[2020-02-10T21:34:56.998Z]     cgroupdriver_systemd_test.go:54: assertion failed: 67108864 (s.HostConfig.Memory int64) != 67108864 (mem int)

[2020-02-10T21:34:56.998Z] FAIL

[2020-02-10T21:34:56.998Z] 

[2020-02-10T21:34:56.998Z] === Failed

[2020-02-10T21:34:56.998Z] === FAIL: arm64.integration.system TestCgroupDriverSystemdMemoryLimit (1.76s)

[2020-02-10T21:34:56.998Z] === PAUSE TestCgroupDriverSystemdMemoryLimit

[2020-02-10T21:34:56.998Z] === CONT  TestCgroupDriverSystemdMemoryLimit

[2020-02-10T21:34:56.998Z]     cgroupdriver_systemd_test.go:54: assertion failed: 67108864 (s.HostConfig.Memory int64) != 67108864 (mem int)
```

AkihiroSuda committed:
TestCgroupDriverSystemdMemoryLimit: fix failure

The test was failing:

  --- FAIL: TestCgroupDriverSystemdMemoryLimit (1.76s)
      cgroupdriver_systemd_test.go:54: assertion failed: 67108864 (s.HostConfig.Memory int64) != 67108864 (mem int)

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
Weird, goimports failing, but I didn't touch these files 🤔 

```
[2020-02-11T12:17:32.067Z] [0;36mINFO[0m Start validation with golang-ci-lint
[2020-02-11T12:17:32.067Z] [36mINFO[0m [config_reader] Used config file hack/validate/golangci-lint.yml 
[2020-02-11T12:17:32.067Z] [36mINFO[0m [lintersdb] Active 14 linters: [deadcode goimports golint gosec gosimple govet ineffassign misspell staticcheck structcheck typecheck unconvert unused varcheck] 
[2020-02-11T12:19:23.506Z] [36mINFO[0m [loader] Go packages loading at mode 575 (files|imports|name|compiled_files|deps|exports_file|types_sizes) took 1m51.398366916s 
[2020-02-11T12:19:24.070Z] [36mINFO[0m [runner/filename_unadjuster] Pre-built 0 adjustments in 452.975549ms 
[2020-02-11T12:21:00.472Z] [36mINFO[0m [runner/max_same_issues] 5/8 issues with text "File is not `goimports`-ed" were hidden, use --max-same-issues 
[2020-02-11T12:21:00.472Z] [36mINFO[0m [runner] Issues before processing: 2333, after processing: 3 
[2020-02-11T12:21:00.472Z] [36mINFO[0m [runner] Processors filtering stat (out/in): exclude-rules: 78/1627, uniq_by_line: 15/15, diff: 15/15, cgo: 1724/2333, filename_unadjuster: 1724/1724, identifier_marker: 1627/1627, exclude: 1627/1627, max_from_linter: 3/3, path_shortener: 3/3, skip_files: 1724/1724, skip_dirs: 1724/1724, max_per_file_from_linter: 8/15, max_same_issues: 3/8, path_prettifier: 1724/1724, autogenerated_exclude: 1627/1724, nolint: 15/78, source_code: 3/3 
[2020-02-11T12:21:00.472Z] [36mINFO[0m [runner] processing took 71.160272ms with stages: nolint: 36.069128ms, identifier_marker: 17.130581ms, path_prettifier: 7.660732ms, exclude-rules: 4.125416ms, autogenerated_exclude: 3.755903ms, skip_dirs: 1.723882ms, cgo: 360.733µs, filename_unadjuster: 200.124µs, source_code: 76.735µs, max_same_issues: 36.294µs, max_per_file_from_linter: 8.107µs, uniq_by_line: 7.791µs, max_from_linter: 2.027µs, path_shortener: 1.524µs, skip_files: 513ns, exclude: 396ns, diff: 386ns 
[2020-02-11T12:21:00.472Z] [36mINFO[0m [runner] linters took 1m35.498244848s with stages: goanalysis_metalinter: 1m18.348317659s, unused: 17.07859655s 
[2020-02-11T12:21:00.472Z] [1mreference/store_test.go:19[0m: [31mFile is not `goimports`-ed[0m (goimports)
[2020-02-11T12:21:00.472Z] 		"registry:5000/foobar:HEAD":                                                        "sha256:470022b8af682154f57a2163d030eb369549549cba00edc69e1b99b46bb924d6",
[2020-02-11T12:21:00.472Z] 		"registry:5000/foobar:alternate":                                                   "sha256:ae300ebc4a4f00693702cfb0a5e0b7bc527b353828dc86ad09fb95c8a681b793",
[2020-02-11T12:21:00.472Z] 		"registry:5000/foobar:latest":                                                      "sha256:6153498b9ac00968d71b66cca4eac37e990b5f9eb50c26877eb8799c8847451b",
[2020-02-11T12:21:00.472Z] 		"registry:5000/foobar:master":                                                      "sha256:6c9917af4c4e05001b346421959d7ea81b6dc9d25718466a37a6add865dfd7fc",
[2020-02-11T12:21:00.472Z] 		"jess/hollywood:latest":                                                            "sha256:ae7a5519a0a55a2d4ef20ddcbd5d0ca0888a1f7ab806acc8e2a27baf46f529fe",
[2020-02-11T12:21:00.472Z] [1mrunconfig/hostconfig_test.go:51[0m: [31mFile is not `goimports`-ed[0m (goimports)
[2020-02-11T12:21:00.472Z] 		"host":           {false, false, true, false, false, false},
[2020-02-11T12:21:00.472Z] 		"container:name": {false, false, false, true, false, false},
[2020-02-11T12:21:00.472Z] 		"none":           {true, false, false, false, true, false},
[2020-02-11T12:21:00.472Z] 		"default":        {true, false, false, false, false, true},
[2020-02-11T12:21:00.472Z] [1mpkg/authorization/api_test.go:20[0m: [31mFile is not `goimports`-ed[0m (goimports)
[2020-02-11T12:21:00.472Z] 		IsCA: true,
[2020-02-11T12:21:00.472Z] [36mINFO[0m File cache stats: 1449 entries of total size 6.8MiB 
[2020-02-11T12:21:00.472Z] [36mINFO[0m Memory: 1550 samples, avg is 273.3MB, max is 3032.7MB 
[2020-02-11T12:21:00.472Z] [36mINFO[0m Execution took 3m27.358511415s               
script returned exit code 1
```

AkihiroSuda commented:
@tianon PTAL?

tianon commented:
I agree with the intent here (automated testing of more things via `systemd`, especially things that can't effectively be tested right now), but anything that runs `systemd` in a container, even more especially one that's typically run with `--privileged`, makes me very wary.  I've seen way too many systems accidentally tanked thanks to `systemd` inside the container having some service someone forgot to disable that then tries to initialize already-initialized host devices (as an example), and causing the expected havoc.

thaJeztah commented:
could we add a variant just used for our CI, and keep the standard dind script as-is?

tianon commented:
I like that idea, as long as the logistics work for actually using it :sweat_smile:

AkihiroSuda commented:
Updated to split the script. PTAL.

AkihiroSuda commented:
@tianon PTAL?

thaJeztah commented on Dockerfile:
```diff
@@ -293,6 +297,10 @@ RUN update-alternatives --set iptables  /usr/sbin/iptables-legacy  || true \
 RUN pip3 install yamllint==1.16.0
 
+# hack/dind switches to hack/dind-systemd when $DOCKER_SYSTEMD=1 is set
+ADD https://raw.githubusercontent.com/AkihiroSuda/containerized-systemd/b70bac0daeea120456764248164c21684ade7d0d/docker-entrypoint.sh /go/src/github.com/docker/docker/hack/dind-systemd
```
Should we add this file to this repository, or are there specific reasons to keep it separate?

AkihiroSuda replied:
Wanted to avoid diverging

cpuguy83 replied:
I'd prefer to do this with `RUN curl` instead, otherwise this will always hit github on build, I think.

AkihiroSuda replied:
updated to use curl

thaJeztah commented on hack/dind:
```diff
@@ -10,6 +10,11 @@ set -e
 
 # Usage: dind CMD [ARG...]
 
+if [ -n "$DOCKER_SYSTEMD" ]; then
```
Wondering if this should be handled through a build-arg, so that `hack/dind` doesn't have to be modified, and an if/else construct in the dockerfile;

```
FROM foo AS systemd-true
COPY hack/dind-systemd hack/dind-systemd
ENTRYPOINT ["hack/dind-systemd"]

FROM foo AS systemd-false
COPY hack/dind-systemd hack/dind
ENTRYPOINT ["hack/dind"]


FROM systemd-${DOCKER_SYSTEMD} AS foo
```

AkihiroSuda replied:
@cpuguy83 @tonistiigi WDYT?

cpuguy83 replied:
I like @thaJeztah's suggestion.

AkihiroSuda replied:
updated

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented on Dockerfile:
```diff
@@ -278,6 +281,8 @@ RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
             python3-setuptools \
             python3-wheel \
+            systemd \
+            systemd-sysv \
```
Should we move the systemd-specific dependencies to the `dev-systemd-true` stage as well?

AkihiroSuda replied:
moved

thaJeztah reviewed (approved):
LGTM, thanks!
